cmd/attest_test.go

```diff
@@ -0,0 +1,279 @@
+package cmd
+
+import (
+	"io"
+	"os"
+	"strconv"
+	"testing"
+
+	"github.com/google/go-tpm-tools/client"
+	"github.com/google/go-tpm-tools/internal/test"
+	"github.com/google/go-tpm/tpm2"
+	"github.com/google/go-tpm/tpmutil"
+)
+
+var getIndex = map[string]uint32{
+	"rsa": client.GceAKTemplateNVIndexRSA,
+	"ecc": client.GceAKTemplateNVIndexECC,
+}
+
+func GCEAKTemplateECC() tpm2.Public {
+	return tpm2.Public{
+		Type:       tpm2.AlgECC,
+		NameAlg:    tpm2.AlgSHA256,
+		Attributes: tpm2.FlagSignerDefault,
+		ECCParameters: &tpm2.ECCParams{
+			Sign: &tpm2.SigScheme{
+				Alg:  tpm2.AlgECDSA,
+				Hash: tpm2.AlgSHA256,
+			},
+			CurveID: 3,
+		},
+	}
+}
+func GCEAKTemplateRSA() tpm2.Public {
+	return tpm2.Public{
+		Type:       tpm2.AlgRSA,
+		NameAlg:    tpm2.AlgSHA256,
+		Attributes: tpm2.FlagSignerDefault,
+		RSAParameters: &tpm2.RSAParams{
+			Sign: &tpm2.SigScheme{
+				Alg:  tpm2.AlgRSASSA,
+				Hash: tpm2.AlgSHA256,
+			},
+			KeyBits: 2048,
+		},
+	}
+}
+
+func setGCETemplate(tb testing.TB, rwc io.ReadWriteCloser, algo string, data []byte) error {
+	var err error
+	idx := tpmutil.Handle(getIndex[algo])
+	if err := tpm2.NVDefineSpace(rwc, tpm2.HandlePlatform, idx,
+		"", "", nil,
+		tpm2.AttrPPWrite|tpm2.AttrPPRead|tpm2.AttrWriteDefine|tpm2.AttrOwnerRead|tpm2.AttrAuthRead|tpm2.AttrPlatformCreate|tpm2.AttrNoDA,
+		uint16(len(data))); err != nil {
+		tb.Fatalf("NVDefineSpace failed: %v", err)
+	}
+	err = tpm2.NVWrite(rwc, tpm2.HandlePlatform, idx, "", data, 0)
+	if err != nil {
+		tb.Fatalf("failed to write NVIndex: %v", err)
+	}
+	return nil
+}
+
+func makeOutputFile(tb testing.TB, output string) string {
+	tb.Helper()
+	file, err := os.CreateTemp("", output)
+	if err != nil {
+		tb.Fatal(err)
+	}
+	defer file.Close()
+	return file.Name()
+}
+
+func TestNonce(t *testing.T) {
+	rwc := test.GetTPM(t)
+	defer client.CheckedClose(t, rwc)
+	ExternalTPM = rwc
+	// Without nonce.
+	RootCmd.SetArgs([]string{"attest", "--key", "AK"})
+	if err := RootCmd.Execute(); err == nil {
+		t.Error("expected not-nil error")
+	}
+	// With odd length nonce.
+	RootCmd.SetArgs([]string{"attest", "--nonce", "12345", "--key", "AK"})
+	if err := RootCmd.Execute(); err == nil {
+		t.Error("expected not-nil error")
+	}
+}
+
+func TestAttestPass(t *testing.T) {
```

Comment:
Any test for SEV-SNP simulation? Note there is a testclient device from go-sev-guest that will select hardware or fake based on the device path.

Reply:
Thank you @deeglaze .
I have added tests for SEV-SNP in attest_test.go .